doc/test-framework/scorecard.md: document --olm-deployed prereq… (#1816)
operator-framework/operator-sdk: doc/test-framework/scorecard.md

**NOTE:** The `ScorecardOutput.Log` field is only intended to be used to log the scorecard's output and the scorecard will ignore that field if a plugin provides it.

## Running the scorecard with an OLM-managed operator

The scorecard can be run using a [Cluster Service Version (CSV)][olm-csv], providing a way to test cluster-ready and non-SDK operators.

Running with a CSV alone requires both the `--csv-path=<CSV manifest path>` and `--olm-deployed` flags to be set. The scorecard assumes your CSV and relevant CRD's have been deployed onto the cluster using OLM when using `--olm-deployed`.

The scorecard requires a proxy container in the operator's `Deployment` pod to read operator logs. A few modifications to your CSV and creation of one extra object are required to run the proxy _before_ deploying your operator with OLM:

> 1. Create a proxy server secret containing a local Kubeconfig:

Reviewer: It shows that required the data from the local Kubeconfig when I understand the following steps are related to the cluster where the OLM files will bee deployed and tested/checked with the score 
```suggestion
1. Create a proxy server secret for your cluster:
```

    1. Generate a username using the scorecard proxy's namespaced owner reference.
          ```sh
          # Substitute "$your_namespace" for the namespace your operator will be deployed in (if any).
          $ echo '{"apiVersion":"","kind":"","name":"scorecard","uid":"","Namespace":"'${your_namespace}'"}' | base64 -w 0
          eyJhcGlWZXJzaW9uIjoiIiwia2luZCI6IiIsIm5hbWUiOiJzY29yZWNhcmQiLCJ1aWQiOiIiLCJOYW1lc3BhY2UiOiJvbG0ifQo=
          ```
    1. Write a `Config` manifest `scorecard-config.yaml` using the following template, substituting `${your_username}` for the base64 username generated above:
          ```yaml
          apiVersion: v1
          kind: Config
          clusters:
          - cluster:
              insecure-skip-tls-verify: true

>               server: http://${your_username}@localhost:8889

Reviewer: Is not here to add the master URL? I mean the URL of the cluster? If yes, IMHO it can lead the user to add the wrong value since will be not always @localhost:8889 or http. WDYT about something as

```suggestion
              server: <CLUSTER_URL>
              
              ....
              
              **NOTE:** CLUSTER_URL is XXX. For example X. 
```

Author: The server is not your cluster URL. This is a scorecard proxy server for the cluster.

Reviewer: So, could we add an explanation over what is the server and how the steps to get this info? WDYT?

Author: We could, but its an implementation detail that is irrelevant to users. Its an internal proxy that the user will never have to worry about.

          ```
    1. Create a `Secret` manifest `scorecard-secret.yaml` containing the operator's namespace (if any) the `Config`'s base64 encoding as a `spec.data` value under the key `kubeconfig`:
          ```yaml
          apiVersion: v1
          kind: Secret
          metadata:
            name: scorecard-kubeconfig
            namespace: ${your_namespace}
          data:
            kubeconfig: ${kubeconfig_base64}
          ```
    1. Apply the secret in-cluster:
          ```sh
          $ kubectl apply -f scorecard-secret.yaml
          ```

>     1. Insert a volume referring to the `Secret` into the operator's `Deployment`:

Reviewer: ```suggestion
    1. Insert a volume referring to the `Secret` into the CSV file in the install section like the following example. 
```

          ```yaml
          spec:
            install:
              spec:
                deployments:
                - name: memcached-operator
                  spec:
                    ...
                    template:
                      ...
                      spec:
                        containers:
                        ...
                        volumes:
                        # scorecard kubeconfig volume
                        - name: scorecard-kubeconfig
                          secret:
                            secretName: scorecard-kubeconfig
                            items:
                            - key: kubeconfig
                              path: config
          ```

> 1. Insert a volume mount and `KUBECONFIG` environment variable into each container in your operator's `Deployment`:

Reviewer: IMHO something as the following can help it be more clear to the users. 
```suggestion
    1. Insert a volume referring to the `Secret` into the CSV file in the install section like the following example. 
```
```suggestion
1. Insert a volume mount and `KUBECONFIG` environment variable into the CSV file in the install section like the following example.  
```

Reviewer: Maybe should be better add all these in the operator.yml and ask to re-run the operator-sdk command to generated the CSV file. WDYT?

    ```yaml
    spec:
      install:
        spec:
          deployments:
          - name: memcached-operator
            spec:
              ...
              template:
                ...
                spec:
                  containers:
                  - name: container1
                    ...
                    volumeMounts:
                    # scorecard kubeconfig volume mount
                    - name: scorecard-kubeconfig
                      mountPath: /scorecard-secret
                    env:
                      # scorecard kubeconfig env
                    - name: KUBECONFIG
                      value: /scorecard-secret/config
                  - name: container2
                    # Do the same for this and all other containers.
                    ...
    ```

> 1. Insert the scorecard proxy container into the operator's `Deployment`:

Reviewer: The same here. IMHO it can be confuse to the user since they can try to look for a "deployment" file. I think is easier to explain to add it in the CSV file and/or operator.yaml

    ```yaml
    spec:
      install:
        spec:
          deployments:
          - name: memcached-operator
            spec:
              ...
              template:
                ...
                spec:
                  containers:
                  ...
                  # scorecard proxy container
                  - name: scorecard-proxy
                    command:
                    - scorecard-proxy
                    env:
                    - name: WATCH_NAMESPACE
                      valueFrom:
                        fieldRef:
                          apiVersion: v1
                          fieldPath: metadata.namespace
                    image: quay.io/operator-framework/scorecard-proxy:master
                    imagePullPolicy: Always
                    ports:
                    - name: proxy
                      containerPort: 8889
    ```

> Alternatively, the [community-operators][community-operators] repo has several bash functions that can perform these operations for you:

Reviewer: Is the following commands,  from the community-operators, doing the same tests that will be done for the above steps? If not, I'd recommend replacing `Alternatively` by something as `Also,`.

```sh
$ curl -Lo csv-manifest-modifiers.sh https://raw.githubusercontent.com/operator-framework/community-operators/master/scripts/lib/file
$ . ./csv-manifest-modifiers.sh
# $NAMESPACE is the namespace your operator will deploy in
$ create_kubeconfig_secret_file scorecard-secret.yaml "$NAMESPACE"
$ kubectl apply -f scorecard-secret.yaml
# $CSV_FILE is the path to your operator's CSV manifest
$ insert_kubeconfig_volume "$CSV_FILE"
$ insert_kubeconfig_secret_mount "$CSV_FILE"
$ insert_proxy_container "$CSV_FILE" "quay.io/operator-framework/scorecard-proxy:master"
```

> Once done, follow the steps in this [document][olm-deploy-operator] to bundle your CSV and CRD's, deploy OLM on minikube or [OKD][okd], and deploy your operator. Once these steps have been completed, run the scorecard with both the `--csv-path=<CSV manifest path>` and `--olm-deployed` flags.

Reviewer: It shows a little confuse to me. Maybe something as : 
```suggestion
After following the steps in this [document][olm-deploy-operator] to bundle your CSV and CRD's and have your operator deployed in the cluster with OLM run the scorecard with both the `--csv-path=<CSV manifest path>` and `--olm-deployed` flags.
```

Author: I prefer the original text, since it clearly states your should follow the steps in the linked document to do everything listed.

**NOTE:** The `ScorecardOutput.Log` field is only intended to be used to log the scorecard's output and the scorecard will ignore that field if a plugin provides it.
To add logs to the main `ScorecardOuput.Log` field, a plugin can output the logs to `stderr`.

> ## Running the scorecard with an OLM-managed operator

Reviewer: nit: required to fix/update the index. I would recommend do it automatically by using doctoc and them removing its comments.

Author: Are we doing this for all documentation now?

Reviewer 2: There was some discussion a while back about how best to integrate doctoc given that it's installed with `npm` (I think). My concern was adding a non-go, non-binary dependency to our build and CI scripts.

Reviewer 2: At the very least in this PR, we can manually fix the DocToc section to match the new header.

Longer-term, @asmacdo is interested in a much larger refactoring and re-homing of our docs to a readthedocs or gitbook site, so I'd rather we not try to make any major changes to our current docs automation (with the exception of @theishshah CLI docs generation)
